lib/trino-parquet/src/main/java/io/trino/parquet/reader/ColumnIndexFilterUtils.java

```diff
@@ -0,0 +1,171 @@
+        {
+            return offsetIndex.getFirstRowIndex(indexMap[pageIndex]);
+        }
+
+        @Override
+        public long getLastRowIndex(int pageIndex, long totalRowCount)
+        {
+            int nextIndex = indexMap[pageIndex] + 1;
+            return (nextIndex >= offsetIndex.getPageCount() ? totalRowCount : offsetIndex.getFirstRowIndex(nextIndex)) - 1;
+        }
+
+        @Override
+        public String toString()
+        {
+            try (Formatter formatter = new Formatter()) {
+                formatter.format("%-12s  %20s  %16s  %20s\n", "", "offset", "compressed size", "first row index");
+                for (int i = 0, n = offsetIndex.getPageCount(); i < n; ++i) {
+                    int index = Arrays.binarySearch(indexMap, i);
+                    boolean isHidden = index < 0;
+                    formatter.format("%spage-%-5d  %20d  %16d  %20d\n",
+                            isHidden ? "- " : "  ",
+                            isHidden ? i : index,
+                            offsetIndex.getOffset(i),
+                            offsetIndex.getCompressedPageSize(i),
+                            offsetIndex.getFirstRowIndex(i));
+                }
+                return formatter.toString();
+            }
+        }
+
+        List<OffsetRange> calculateOffsetRanges(long rowGroupOffset)
+        {
+            List<OffsetRange> ranges = new ArrayList<>();
+            int n = this.getPageCount();
+            if (n > 0) {
+                OffsetRange currentRange = null;
+
+                long firstPageOffset = this.offsetIndex.getOffset(0);
+                // Add a range for the dictionary page if required
+                if (rowGroupOffset < firstPageOffset) {
+                    // We need to adjust the offset by startingPosition for presto because dataSource.readFully() started at startingPosition
+                    currentRange = new OffsetRange(rowGroupOffset /*- startingPosition*/, (int) (firstPageOffset - rowGroupOffset));
+                    ranges.add(currentRange);
+                }
+
+                for (int i = 0; i < n; ++i) {
+                    long offset = this.getOffset(i);
+                    int length = this.getCompressedPageSize(i);
+                    // We need to adjust the offset by startingPosition for presto because dataSource.readFully() started at startingPosition
+                    if (currentRange == null || !currentRange.extend(offset /*- startingPosition*/, length)) {
+                        currentRange = new OffsetRange(offset /*- startingPosition*/, length);
+                        ranges.add(currentRange);
+                    }
+                }
+            }
+            return ranges;
+        }
+    }
+
+    /*
+     * Returns the filtered offset index containing only the pages which are overlapping with rowRanges.
+     */
+    static FilteredOffsetIndex filterOffsetIndex(OffsetIndex offsetIndex, RowRanges rowRanges, long totalRowCount)
```

Comment:
This could be a method of `OffsetIndex`, which could be a top-level class (making `ColumnIndexFilterUtils` unnecessary).